# 09 Portfolio Holdings - Data Cleaning

## Objective

Clean and preprocess the Portfolio Holdings dataset before using it for Exploratory Data Analysis (EDA).

The cleaning process includes:

- Loading the raw dataset
- Inspecting dataset structure
- Checking missing values
- Identifying duplicate records
- Correcting data types
- Validating cleaned data
- Exporting the cleaned dataset

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

In [2]:
holding_df = pd.read_csv("../Data/Raw/09_portfolio_holdings.csv")

print("DATASET INFORMATION")

display(holding_df.head())

print("\nShape:", holding_df.shape)

print("\nColumns:")
print(holding_df.columns)

print("\nData Types:")
print(holding_df.dtypes)

print("\nMissing Values:")
display(holding_df.isnull().sum())

print("\nDuplicate Records:")
print(holding_df.duplicated().sum())

DATASET INFORMATION


,amfi_code,stock_symbol,stock_name,sector,weight_pct,market_value_cr,current_price_inr,portfolio_date
0,119551,POWERGRID,Power Grid Corporation,Utilities,13.85,737.09,6011.08,2025-12-31
1,119551,HDFCBANK,HDFC Bank Ltd,Banking,11.19,88.97,1074.65,2025-12-31
2,119551,GRASIM,Grasim Industries Ltd,Diversified,9.90,208.45,5964.59,2025-12-31
3,119551,DRREDDY,Dr. Reddy's Laboratories,Pharma,4.76,161.32,3748.82,2025-12-31
4,119551,ASIANPAINT,Asian Paints Ltd,Paints,10.25,725.90,1321.45,2025-12-31



Shape: (322, 8)

Columns:
Index(['amfi_code', 'stock_symbol', 'stock_name', 'sector', 'weight_pct',
       'market_value_cr', 'current_price_inr', 'portfolio_date'],
      dtype='object')

Data Types:
amfi_code              int64
stock_symbol          object
stock_name            object
sector                object
weight_pct           float64
market_value_cr      float64
current_price_inr    float64
portfolio_date        object
dtype: object

Missing Values:


amfi_code            0
stock_symbol         0
stock_name           0
sector               0
weight_pct           0
market_value_cr      0
current_price_inr    0
portfolio_date       0
dtype: int64


Duplicate Records:
0


In [4]:
# Convert portfolio date to datetime
holding_df["portfolio_date"] = pd.to_datetime(holding_df["portfolio_date"])

# Sort records
holding_df = holding_df.sort_values(
    ["amfi_code", "weight_pct"],
    ascending=[True, False]
).reset_index(drop=True)

# Verify
holding_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322 entries, 0 to 321
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   amfi_code          322 non-null    int64         
 1   stock_symbol       322 non-null    object        
 2   stock_name         322 non-null    object        
 3   sector             322 non-null    object        
 4   weight_pct         322 non-null    float64       
 5   market_value_cr    322 non-null    float64       
 6   current_price_inr  322 non-null    float64       
 7   portfolio_date     322 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float64(3), int64(1), object(3)
memory usage: 20.2+ KB


In [5]:
print("DATA VALIDATION")

print(f"Shape : {holding_df.shape}")
print(f"Duplicate Rows : {holding_df.duplicated().sum()}")

print("\nMissing Values")
print(holding_df.isnull().sum())

print("\nPortfolio Date")
print(holding_df["portfolio_date"].min())
print(holding_df["portfolio_date"].max())

print("\nUnique Schemes :", holding_df["amfi_code"].nunique())

print("Unique Sectors :", holding_df["sector"].nunique())

DATA VALIDATION
Shape : (322, 8)
Duplicate Rows : 0

Missing Values
amfi_code            0
stock_symbol         0
stock_name           0
sector               0
weight_pct           0
market_value_cr      0
current_price_inr    0
portfolio_date       0
dtype: int64

Portfolio Date
2025-12-31 00:00:00
2025-12-31 00:00:00

Unique Schemes : 34
Unique Sectors : 14


In [6]:
output_path = "../Data/Processed/09_portfolio_holdings_cleaned.csv"

holding_df.to_csv(output_path, index=False)

print(f"Cleaned dataset saved to:\n{output_path}")

Cleaned dataset saved to:
../Data/Processed/09_portfolio_holdings_cleaned.csv


## Cleaning Summary

### Cleaning Performed

- Converted the `portfolio_date` column from object to datetime format.
- Sorted records by AMFI code and portfolio weight.
- Verified duplicate records.
- Verified missing values.
- Exported the cleaned dataset to the Processed folder.

### Observations

- No missing values were identified.
- No duplicate records were present.
- Only the `portfolio_date` column required data type conversion.

### Final Dataset

- Records: 322
- Features: 8
- Portfolio Date: 31 December 2025